kernel/src/cpu/apic.rs

```diff
@@ -0,0 +1,851 @@
+            _ => Self::Dest,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum IcrMessageType {
+    Fixed = 0,
+    Unknown = 3,
+    Nmi = 4,
+    Init = 5,
+    Sipi = 6,
+    ExtInt = 7,
+}
+
+impl IcrMessageType {
+    const fn into_bits(self) -> u64 {
+        self as _
+    }
+    const fn from_bits(value: u64) -> Self {
+        match value {
+            7 => Self::ExtInt,
+            6 => Self::Sipi,
+            5 => Self::Init,
+            4 => Self::Nmi,
+            0 => Self::Fixed,
+            _ => Self::Unknown,
+        }
+    }
+}
+
+#[bitfield(u64)]
+struct ApicIcr {
+    pub vector: u8,
+    #[bits(3)]
+    pub message_type: IcrMessageType,
+    pub destination_mode: bool,
+    pub delivery_status: bool,
+    rsvd_13: bool,
+    pub assert: bool,
+    pub trigger_mode: bool,
+    #[bits(2)]
+    pub remote_read_status: usize,
+    #[bits(2)]
+    pub destination_shorthand: IcrDestFmt,
+    #[bits(12)]
+    rsvd_31_20: u64,
+    pub destination: u32,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum ApicError {
+    ApicError,
+}
+
+#[derive(Default, Clone, Copy, Debug)]
+pub struct LocalApic {
+    irr: [u32; 8],
+    allowed_irr: [u32; 8],
+    isr_stack_index: usize,
+    isr_stack: [u8; 16],
+    tmr: [u32; 8],
+    host_tmr: [u32; 8],
+    update_required: bool,
+    interrupt_delivered: bool,
+    interrupt_queued: bool,
+    lazy_eoi_pending: bool,
+    nmi_pending: bool,
+}
+
+impl LocalApic {
+    pub fn new() -> Self {
+        LocalApic {
+            irr: [0; 8],
+            allowed_irr: [0; 8],
+            isr_stack_index: 0,
+            isr_stack: [0; 16],
+            tmr: [0; 8],
+            host_tmr: [0; 8],
+            update_required: false,
+            interrupt_delivered: false,
+            interrupt_queued: false,
+            lazy_eoi_pending: false,
+            nmi_pending: false,
+        }
+    }
+
+    fn scan_irr(&self) -> u8 {
+        // Scan to find the highest pending IRR vector.
+        for i in (0..7).rev() {
+            if self.irr[i] != 0 {
+                let bit_index = 31 - self.irr[i].leading_zeros();
+                let vector = (i as u32) * 32 + bit_index;
+                return vector.try_into().unwrap();
+            }
+        }
+        0
+    }
+
+    fn remove_vector_register(register: &mut [u32; 8], irq: u8) {
+        register[irq as usize >> 5] &= !(1 << (irq & 31));
+    }
+
+    fn insert_vector_register(register: &mut [u32; 8], irq: u8) {
+        register[irq as usize >> 5] |= 1 << (irq & 31);
+    }
+
+    fn test_vector_register(register: &[u32; 8], irq: u8) -> bool {
+        (register[irq as usize >> 5] & 1 << (irq & 31)) != 0
+    }
+
+    fn rewind_pending_interrupt(&mut self, irq: u8) {
+        assert!(self.isr_stack_index != 0);
+        assert!(self.isr_stack[self.isr_stack_index - 1] == irq);
+        Self::insert_vector_register(&mut self.irr, irq);
+        self.isr_stack_index -= 1;
+        self.update_required = true;
```

Comment:
Using `checked_sub()` (and using a safe array accessor) is probably more readable:

```suggestion
        let stack_index = self.isr_stack_index.checked_sub(1).unwrap();
        assert!(self.isr_stack.get(stack_index) == Some(irq));
        Self::insert_vector_register(&mut self.irr, irq);
        self.isr_stack_index = stack_index;
        self.update_required = true;
```

Although I'd suggest returning an error.

Reply:
An error here is not appropriate, because the state of the ISR stack is not subject to runtime conditions, i.e. its safety does not depend on set of interrupts in service.  The architecture of APIC interrupt priorities guarantees that it is never possible to have more then 15 interrupts in service at a time.  If any of the assumptions here are violated, it is indicative of a fatal bug in APIC emulation, which is certainly not an error that any caller can be expected to deal with.

Using `checked_sub()` seems advisable because it will panic in case of a bug-induced error instead of an underflow that would cause some other subtle problem elsewhere, so I will make that change.

I don't agree that the `get(i)` notation is more readable than `[i]` notation, particularly in cases where the index is guaranteed to exist.  Particularly since this is only present in a debug assertion, it's not particularly helpful to guard against unexpected bounds violations - in fact, I would prefer to see a panic in such a case to help illustrate that the assertion has failed.

Reply:
The reason for using `.get(i)` is that it won't immediately panic if `i` is out of bounds, not readability. I would then suggest using `assert_eq!(self.isr_stack.get(stack_index), Some(irq))` so that the panic message displays both the obtained value (which will be `None` on an out of bounds access or `Some(val)` if `val` is not expected) and the expected value.

Reply:
I see - you're saying that using `get` means that the assertion failure is more readable because it actually displays as an assertion failure instead of just getting a random panic.  I'm not sure it makes a significant difference when debugging, but I do observe that there is no other assert in the code base that contains an expression that is subject to panic.  Therefore, following your suggestion would be the most consistent so we can maintain the pattern of no panics within assertion clauses.